### Importing Required Libraries

In [1]:
import tensorflow as tf
import keras

Using TensorFlow backend.


In [2]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Flatten, Dense

In [3]:
import numpy as np
import pandas as pd

### Reading the file using Pandas

In [4]:
df = pd.read_csv('Churn_Modelling.csv')
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
# Checking Null Values
df.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

In [6]:
# Finding dtype of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [7]:
# Separating Target Variable and dropping not required columns.
X = df.drop(labels=['CustomerId', 'Surname', 'RowNumber', 'Exited'], axis = 1)
y = df['Exited']

In [8]:
print(X.shape, y.shape)

(10000, 10) (10000,)


### Label Encoding

#### I'm taking this is an optional

In [9]:
#from sklearn.preprocessing import LabelEncoder
#label = LabelEncoder()

In [10]:
#X['Geography'] = label.fit_transform(X['Geography'])

In [11]:
#X['Gender'] = label.fit_transform(X['Gender'])
#X.head()

### Getting Dummies

In [12]:
X = pd.get_dummies(X, drop_first=True, columns=['Geography', 'Gender'])
X.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88                  0                0   
1               1        112542.58                  0                1   
2               0        113931.57                  0                0   
3               0         93826.63                  0                0   
4               1         79084.10                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0

### Train Test Split

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

In [14]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [15]:
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [16]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(8000, 11) (8000,)
(2000, 11) (2000,)


### Building ANN

In [17]:
model = Sequential()
model.add(Flatten())
model.add(Dense(X.shape[1], activation='relu'))
model.add(Dense(128, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))

#### 1) Optimizer is Adam

In [18]:
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

In [19]:
model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 2s 269us/sample - loss: 0.4975 - accuracy: 0.7856
Epoch 2/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.4199 - accuracy: 0.8217
Epoch 3/10
8000/8000 [==============================] - 1s 73us/sample - loss: 0.3913 - accuracy: 0.8391
Epoch 4/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3676 - accuracy: 0.8475
Epoch 5/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3560 - accuracy: 0.8539
Epoch 6/10
8000/8000 [==============================] - 1s 73us/sample - loss: 0.3492 - accuracy: 0.8559
Epoch 7/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3453 - accuracy: 0.8584
Epoch 8/10
8000/8000 [==============================] - 1s 78us/sample - loss: 0.3416 - accuracy: 0.8601
Epoch 9/10
8000/8000 [==============================] - 1s 75us/sample - loss: 0.3403 - accuracy: 0.8610
Epoch 10/10
8000/8000 [=========

In [20]:
prediction = model.predict_classes(X_test)

In [21]:
model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.34102217948436736, 0.856]

In [22]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

In [23]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.856
Precision Score:  0.6778115501519757
Recall Score:  0.5506172839506173
F1 Score:  0.6076294277929156
Confusion Matrix:
 [[1489  106]
 [ 182  223]]


#### 2) Optimizer is Adagrad

In [24]:
model.compile(optimizer='adagrad', loss = 'binary_crossentropy', metrics=['accuracy'])

In [25]:
model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 2s 190us/sample - loss: 0.3339 - accuracy: 0.8640
Epoch 2/10
8000/8000 [==============================] - 1s 73us/sample - loss: 0.3328 - accuracy: 0.8636
Epoch 3/10
8000/8000 [==============================] - 1s 72us/sample - loss: 0.3326 - accuracy: 0.8644
Epoch 4/10
8000/8000 [==============================] - 1s 80us/sample - loss: 0.3324 - accuracy: 0.8641
Epoch 5/10
8000/8000 [==============================] - 1s 73us/sample - loss: 0.3323 - accuracy: 0.8645
Epoch 6/10
8000/8000 [==============================] - 1s 72us/sample - loss: 0.3322 - accuracy: 0.8649
Epoch 7/10
8000/8000 [==============================] - 1s 76us/sample - loss: 0.3322 - accuracy: 0.8640
Epoch 8/10
8000/8000 [==============================] - 1s 72us/sample - loss: 0.3321 - accuracy: 0.8648
Epoch 9/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3321 - accuracy: 0.8650
Epoch 10/10
8000/8000 [=========

In [26]:
prediction = model.predict_classes(X_test)

In [27]:
model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.33432381176948545, 0.858]

In [28]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.858
Precision Score:  0.7168458781362007
Recall Score:  0.49382716049382713
F1 Score:  0.5847953216374268
Confusion Matrix:
 [[1516   79]
 [ 205  200]]


#### 3) Optimizer is SGD(Stochastic Gradient Descent)

In [29]:
model.compile(optimizer='SGD', loss = 'binary_crossentropy', metrics=['accuracy'])

In [30]:
model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 1s 165us/sample - loss: 0.3329 - accuracy: 0.8633
Epoch 2/10
8000/8000 [==============================] - 1s 80us/sample - loss: 0.3326 - accuracy: 0.8634
Epoch 3/10
8000/8000 [==============================] - 1s 75us/sample - loss: 0.3325 - accuracy: 0.8646
Epoch 4/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3322 - accuracy: 0.8648
Epoch 5/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3320 - accuracy: 0.8644
Epoch 6/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3316 - accuracy: 0.8648
Epoch 7/10
8000/8000 [==============================] - 1s 75us/sample - loss: 0.3317 - accuracy: 0.8631
Epoch 8/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3315 - accuracy: 0.8651
Epoch 9/10
8000/8000 [==============================] - 1s 77us/sample - loss: 0.3313 - accuracy: 0.8629
Epoch 10/10
8000/8000 [=========

In [31]:
prediction = model.predict_classes(X_test)

In [32]:
model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.333136519908905, 0.8615]

In [33]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.8615
Precision Score:  0.7269503546099291
Recall Score:  0.5061728395061729
F1 Score:  0.596797671033479
Confusion Matrix:
 [[1518   77]
 [ 200  205]]


#### 4) Optimizer is RMSprop

In [34]:
model.compile(optimizer='RMSprop', loss = 'binary_crossentropy', metrics=['accuracy'])

In [35]:
model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 2s 235us/sample - loss: 0.3350 - accuracy: 0.8619
Epoch 2/10
8000/8000 [==============================] - 1s 82us/sample - loss: 0.3336 - accuracy: 0.8637
Epoch 3/10
8000/8000 [==============================] - 1s 83us/sample - loss: 0.3327 - accuracy: 0.8644
Epoch 4/10
8000/8000 [==============================] - 1s 86us/sample - loss: 0.3326 - accuracy: 0.8635
Epoch 5/10
8000/8000 [==============================] - 1s 75us/sample - loss: 0.3311 - accuracy: 0.8630
Epoch 6/10
8000/8000 [==============================] - 1s 75us/sample - loss: 0.3311 - accuracy: 0.8635
Epoch 7/10
8000/8000 [==============================] - 1s 77us/sample - loss: 0.3302 - accuracy: 0.8654
Epoch 8/10
8000/8000 [==============================] - 1s 70us/sample - loss: 0.3288 - accuracy: 0.8641
Epoch 9/10
8000/8000 [==============================] - 1s 74us/sample - loss: 0.3283 - accuracy: 0.8634
Epoch 10/10
8000/8000 [=========

In [36]:
prediction = model.predict_classes(X_test)

In [37]:
model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.33438518905639647, 0.86]

In [38]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.86
Precision Score:  0.7323420074349443
Recall Score:  0.48641975308641977
F1 Score:  0.5845697329376854
Confusion Matrix:
 [[1523   72]
 [ 208  197]]


### By Comparing both accuaracy score and F1 score we can clearly identify that the model is over fitting
### To eliminate the over fitting we use dropout in the hidden layers

Dropout is a regularization technique patented by Google for reducing overfitting in neural networks by preventing complex co-adaptations on training data. It is a very efficient way of performing model averaging with neural networks. The term "dropout" refers to dropping out units (both hidden and visible) in a neural network

In [39]:
# Importing the Dropout()
from tensorflow.keras.layers import Dropout

### Activation Relu

In [40]:
D_model = Sequential()
D_model.add(Dropout(0.2))
D_model.add(Flatten())
D_model.add(Dropout(0.5))
D_model.add(Dense(X.shape[1], activation='relu'))
D_model.add(Dropout(0.5))
D_model.add(Dense(128, activation='relu'))
D_model.add(Dropout(0.5))
D_model.add(Dense(1, activation = 'sigmoid'))

In [41]:
D_model.compile(optimizer='SGD', loss = 'binary_crossentropy', metrics=['accuracy'])

In [42]:
D_model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 2s 221us/sample - loss: 0.6014 - accuracy: 0.7549
Epoch 2/10
8000/8000 [==============================] - 1s 91us/sample - loss: 0.5462 - accuracy: 0.7935
Epoch 3/10
8000/8000 [==============================] - 1s 82us/sample - loss: 0.5361 - accuracy: 0.7908
Epoch 4/10
8000/8000 [==============================] - 1s 81us/sample - loss: 0.5234 - accuracy: 0.7944
Epoch 5/10
8000/8000 [==============================] - 1s 85us/sample - loss: 0.5176 - accuracy: 0.7939
Epoch 6/10
8000/8000 [==============================] - 1s 88us/sample - loss: 0.5158 - accuracy: 0.7940
Epoch 7/10
8000/8000 [==============================] - 1s 85us/sample - loss: 0.5143 - accuracy: 0.7937
Epoch 8/10
8000/8000 [==============================] - 1s 87us/sample - loss: 0.5077 - accuracy: 0.7937
Epoch 9/10
8000/8000 [==============================] - 1s 82us/sample - loss: 0.5062 - accuracy: 0.7952
Epoch 10/10
8000/8000 [=========

In [43]:
prediction = D_model.predict_classes(X_test)

In [44]:
D_model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.48650478529930113, 0.7975]

In [45]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.7975
Precision Score:  0.0
Recall Score:  0.0
F1 Score:  0.0
Confusion Matrix:
 [[1595    0]
 [ 405    0]]


C:\Users\User\Anaconda3\envs\tf\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Activation is Softmax

In [46]:
D_model = Sequential()
D_model.add(Flatten())
D_model.add(Dense(X.shape[1], activation='softmax'))
D_model.add(Dropout(0.5))
D_model.add(Dense(128, activation='softmax'))
D_model.add(Dropout(0.5))
D_model.add(Dense(1, activation = 'sigmoid'))

In [47]:
D_model.compile(optimizer='SGD', loss = 'binary_crossentropy', metrics=['accuracy'])

In [48]:
D_model.fit(X_train, y_train.to_numpy(),epochs = 10)

Train on 8000 samples
Epoch 1/10
8000/8000 [==============================] - 2s 197us/sample - loss: 0.6172 - accuracy: 0.7925
Epoch 2/10
8000/8000 [==============================] - 1s 82us/sample - loss: 0.5436 - accuracy: 0.7960
Epoch 3/10
8000/8000 [==============================] - 1s 81us/sample - loss: 0.5199 - accuracy: 0.7960
Epoch 4/10
8000/8000 [==============================] - 1s 80us/sample - loss: 0.5115 - accuracy: 0.7960
Epoch 5/10
8000/8000 [==============================] - 1s 83us/sample - loss: 0.5083 - accuracy: 0.7960
Epoch 6/10
8000/8000 [==============================] - 1s 83us/sample - loss: 0.5068 - accuracy: 0.7960
Epoch 7/10
8000/8000 [==============================] - 1s 82us/sample - loss: 0.5064 - accuracy: 0.7960
Epoch 8/10
8000/8000 [==============================] - 1s 80us/sample - loss: 0.5061 - accuracy: 0.7960
Epoch 9/10
8000/8000 [==============================] - 1s 91us/sample - loss: 0.5060 - accuracy: 0.7960
Epoch 10/10
8000/8000 [=========

In [49]:
prediction = D_model.predict_classes(X_test)

In [50]:
D_model.evaluate(X_test, y_test.to_numpy())

2000/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.503883594751358, 0.7975]

In [51]:
print("Accuracy Score: ", accuracy_score(y_test, prediction))
print("Precision Score: ", precision_score(y_test, prediction))
print("Recall Score: ", recall_score(y_test, prediction))
print("F1 Score: ", f1_score(y_test, prediction))
print("Confusion Matrix:\n", confusion_matrix(y_test, prediction))

Accuracy Score:  0.7975
Precision Score:  0.0
Recall Score:  0.0
F1 Score:  0.0
Confusion Matrix:
 [[1595    0]
 [ 405    0]]


C:\Users\User\Anaconda3\envs\tf\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### By observing the accuarcy scores of the models before and after using Dropout.
#### Before using Dropout: 86% (appx.)
#### After using Droput: 80%(appx.)

### There is no huge difference in Accuracy score after using the Dropout in the model.